components/Wallet/Replace.tsx

```diff
@@ -0,0 +1,245 @@
+  const wallet = useWallet()
+  const cid = router.query.cid as string
+  const { loginOption, walletProvider, walletError, resetWalletError } =
+    useWalletProvider()
+
+  // Input states
+  const [expert, setExpert] = useState<boolean>(false)
+  const [gasPremium, setGasPremium] = useState<FilecoinNumber | null>(null)
+  const [gasLimit, setGasLimit] = useState<FilecoinNumber | null>(null)
+  const [gasFeeCap, setGasFeeCap] = useState<FilecoinNumber | null>(null)
+  const [isGasPremiumValid, setIsGasPremiumValid] = useState<boolean>(false)
+  const [isGasLimitValid, setIsGasLimitValid] = useState<boolean>(false)
+  const [isGasFeeCapValid, setIsGasFeeCapValid] = useState<boolean>(false)
+  const inputsValid = isGasPremiumValid && isGasLimitValid && isGasFeeCapValid
+
+  // Sending states
+  const [isSending, setIsSending] = useState<boolean>(false)
+  const [sendError, setSendError] = useState<Error | null>(null)
+
+  // Load data
+  const {
+    message,
+    loading: messageLoading,
+    error: messageError
+  } = useGetMessage(cid)
+  const {
+    gasParams,
+    loading: gasParamsLoading,
+    error: gasParamsError
+  } = useGetReplaceMessageGasParams(walletProvider, message, false)
+  const {
+    gasParams: minGasParams,
+    loading: minGasParamsLoading,
+    error: minGasParamsError
+  } = useGetReplaceMessageGasParams(walletProvider, message, true)
+
+  // Data states
+  const hasLoadError = !!(messageError || gasParamsError || minGasParamsError)
+  const isLoading = messageLoading || gasParamsLoading || minGasParamsLoading
+  const isLoaded = !!(message && gasParams && minGasParams)
+
+  // Calculate maximum transaction fee
+  const maxFee = useMemo<FilecoinNumber | null>(() => {
+    return gasLimit && gasFeeCap ? getMaxGasFee(gasFeeCap, gasLimit) : null
+  }, [gasLimit, gasFeeCap])
+
+  // Set default gas params after loading
+  // the data or disabling expert mode
+  useEffect(() => {
+    if (!expert && gasParams) {
+      setGasPremium(gasParams.gasPremium)
+      setGasLimit(gasParams.gasLimit)
+      setGasFeeCap(gasParams.gasFeeCap)
+    }
+  }, [expert, gasParams])
+
+  // Attempt sending message
+  const onSend = async () => {
+    setIsSending(true)
+    setSendError(null)
+    resetWalletError()
+    const replace = strategy === ReplaceStrategy.CANCEL
+    const newMessage = new Message({
+      to: message.to,
+      from: message.from,
+      nonce: message.nonce,
+      value: replace ? 0 : message.value,
+      method: replace ? 0 : message.method,
+      params: replace ? '' : message.params,
+      gasPremium: gasPremium.toAttoFil(),
+      gasFeeCap: gasFeeCap.toAttoFil(),
+      gasLimit: new BigNumber(gasLimit.toAttoFil()).toNumber()
+    })
+    try {
+      const lotusMessage = newMessage.toLotusType()
+      const signedMessage = await walletProvider.wallet.sign(
+        wallet.address,
+        lotusMessage
+      )
+      const msgValid = await walletProvider.simulateMessage(lotusMessage)
```

Comment:
I wonder if this isn't catching cases where the original message has already confirmed? We might need another call to see if our original message has already made it into a block